javascripts/content_scripts/privly.js

```diff
@@ -602,10 +613,169 @@ var privly = {
+
+    if (iframe === undefined) {
+      return;
+    }
+
+    //hides the iframe if it's not already hidden
+    if (iframe.getAttribute("data-privly-display") === "true") {
+        iframe.setAttribute("data-privly-display", "false");
+        iframe.style.display = "none";
+    }
+
+    //finds the link and displays it
+    var link = iframe.nextSibling;
+    if(link.getAttribute("data-privlyhref")==url){
+      //check if it is not already displayed
+      if(link.getAttribute("data-privly-display")=="false"){
+        link.setAttribute("data-privly-display","true");
+        link.style.display = "inherit";
+      }
+    }
+
+  },
+
+  /**
+   * Receive an iframe show message sent by the iframe using postMessage.
+   * This function receives a message containing the show command, and
+   * displays the hidden iframe accordingly. The URL appearing at the place while
+   * iframe was hidden gets replaced by the iframe. This is done by hiding the url.
+   *
+   * @param {message} message A posted message in JSON format from one of the 
+   * trusted domains. It contains the command, the name of the iframe, and 
+   * height of the iframe(optional).
+   * contents. Ex: {"command":"show", "frameID":"ifrm0"}
+   *
+   */
+  showIframePostedMessage: function(message){
+
+    "use strict";
+
+    //check the format of the message
+     if (typeof(message.origin) !== "string"){
+        return;
+    }
+
+    var msg = JSON.parse(message.data);
+
+    if(msg.command != 'show'){
+      return;h
+    }
+
+    //Get the element by name.
+    var frameName = msg.frameID;
+    var iframe = document.getElementsByName(frameName)[0];
+    var url = iframe.getAttribute("data-privlyHref");
+
+    if (iframe === undefined) {
+      return;
+    }
+
+    //shows the iframe
+    if (iframe.getAttribute("data-privly-display") === "false") {
+        iframe.setAttribute("data-privly-display", "true");
+        iframe.style.display = "";
+    }
+
+    //finds the link and hides it
+    var link = iframe.nextSibling;
+    if(link.getAttribute("data-privlyhref")==url){
+      if(link.getAttribute("data-privly-display")=="true"){
+        link.setAttribute("data-privly-display","false");
+        link.style.display = "none";
+      }
     }
+
   },
 
+
+  /**
+   * Receive an iframe remove message sent by the iframe using postMessage.
+   * This function receives a message containing the remove command and name of the iframe, 
+   * and removes the iframe accordingly. After removing it replaces iframe with the
+   * privly URL.
+   *
+   * @param {message} message A posted message in JSON format from one of the 
+   * trusted domains. It contains the command, the name of the iframe, and 
+   * height of the iframe(optional).
+   * contents. Ex: {"command":"remove", "frameID":"ifrm0"}
+   *
+   */
+  removeIframePostedMessage: function(message){
+
+    "use strict";
+
+    //check the format of the message
+     if (typeof(message.origin) !== "string"){
+        return;
```

Comment:
You repeat the code for checking the message format several times and call all the message listeners every time a message is received. Define a single message listener and have it dispatch to the proper message handler depending on the contents of the message.